### PR TITLE
Support skipping keyUsage check before signing

### PR DESCRIPTION
**Pull request by airtower-luna**
This is useful for tests, both of xades4j itself and other software. The tests for this feature (in `UncheckedSignerBESTest.java`) unfortunately demonstrate that such tests are useful, because I found the current xades4j *verification* procedure doesn't check the keyUsage of the signing certificate (see #230).

Before merging I'd like to reduce code duplication by making the `xades4j.production.Unchecked*` classes inherit from their regular counterparts, but I'd like to confirm first that the approach is acceptable for merging in general. `KeyInfoBuilder.buildKeyInfo(...)` would have to be split into two methods internally, so the `UncheckedKeyInfoBuilder` could skip the keyUsage validation part, and I'd like to make `SignerBES.keyInfoBuilder` protected so the uncheck version could overwrite it with an `UncheckedKeyInfoBuilder`.

### CODE REVIEW

**Comment by airtower-luna** [comment]
Rebased on current `master`.

**Comment by luisgoncalves** [comment]
Thanks for your interest in xades4j!

I think the proposed approach is not scalable. If another "variation" similar to "don't check key usage" was to be added, would that also be a new subclass of the profile and/or `KeyInfoBuilder`? Which would extend which?

I think a better way to implement this would be to add a new option to `BasicSignatureOptions`. Default value should preserve current behavior. What do you think?

**Comment by airtower-luna** [comment]
> I think a better way to implement this would be to add a new option to BasicSignatureOptions. Default value should preserve current behavior, i.e. enfore the check. What do you think?

You're right, that's a much better approach. I've reworked the code in that way, and it's a lot simpler. :slightly_smiling_face:

File: src/main/java/xades4j/production/BasicSignatureOptions.java [hunk 1]
```diff
@@ -24,12 +24,33 @@
...
     private boolean includePublicKey;
     private boolean signKeyInfo;
 
+    /**
+     * Configures whether to check that the keyUsage of the signing certificate
+     * allows use for signing before creating a signature. If enabled (the default)
+     * signing will fail if the keyUsage of the certificate does not allow signing.
+     * You should only disable this for testing.
+     *
+     * @param enabled {@code true} to enable the check, {@code false to disable}
```
Comment (luisgoncalves) [inline src/main/java/xades4j/production/BasicSignatureOptions.java:40]: Detail: "to disable" should be outside the code tag

**Comment by luisgoncalves** [comment]
This looks good, thanks!

But there seems to be some problem with the keystore used for tests, as the tests are failing (both on CI and on my machine).

File: src/main/java/xades4j/production/BasicSignatureOptions.java (hunk 1 above)
Comment (airtower-luna) [reply src/main/java/xades4j/production/BasicSignatureOptions.java]: Fixed, thank you.

**Comment by airtower-luna** [comment]
I created the keystore with OpenJDK 11 from Debian Sid, I wonder if there's some format incompatibility? On an internal CI server with Ubuntu 18.04 I've seen a keystore issue too, on one with Ubuntu 20.04 it works.

Anyway, I've removed the pre-generated keystore and changed the tests to load the test CAs directly in 9966cc1ba26ba8aa6a01a6c0b4e073aba8d100ba, I hope that works reliably across Java versions. :slightly_smiling_face:

**Comment by luisgoncalves** [comment]
Is it possible that you forgot to push the changes to the tests? :)

**Comment by airtower-luna** [comment]
Looks like I did and just pushed the keystore deletion, sorry. Should be fixed now. :sweat_smile:

**Comment by luisgoncalves** [comment]
Thanks for the contribution

**Pull request merged**